### PR TITLE
csound: build on Linux

### PR DESCRIPTION
**Pull request by nwhetsell**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is an attempt to build Csound on Linux. A few things worth pointing out:

* On macOS, Csound depends on [WiiUse](https://github.com/wiiuse/wiiuse), which [does not currently have a Linux bottle](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/wiiuse.rb). I made Csound depend on WiiUse only on macOS. (It *is* possible to build WiiUse on Linux, but I had to do `sudo apt-get install libbluetooth-dev` first, and I’m not sure how to encapsulate that into Homebrew.)

* Csound’s Java bindings work on Linux (locally, at least), but this makes the `caveats` a bit complicated (and also requires, as best as I can determine, having `sudo` commands in the `caveats` on Linux). I’m not sure if this is acceptable.

### CODE REVIEW

**Comment by nwhetsell** [comment]
The test failure is a Csound segmentation fault: https://github.com/Homebrew/homebrew-core/runs/6429886372?check_suite_focus=true#step:7:71. I don’t have a way to debug this, unfortunately.

**Comment by nwhetsell** [comment]
In an attempt to figure this out, I duplicated the Csound formula in a tap at https://github.com/nwhetsell/homebrew-csound/tree/csound-linux. Weirdly, the Csound tests pass on Linux in this tap:

https://github.com/nwhetsell/homebrew-csound/runs/6435173701?check_suite_focus=true#step:8:41

It seems like there’s a difference between `ubuntu-latest` configurations that results in a Csound that segfaults immediately. One difference I can see is in the output of `brew config`. [Here at homebrew-core](https://github.com/Homebrew/homebrew-core/runs/6431241324?check_suite_focus=true#step:6:41), there’s:

<pre>
CPU: dual-core 64-bit haswell
<strong>Clang: N/A</strong>
Git: 2.36.1 => /usr/bin/git
Curl: 7.47.0 => /usr/bin/curl
Kernel: Linux 5.13.0-1022-azure x86_64 GNU/Linux
OS: Unknown
<strong>Host glibc: 2.23</strong>
<strong>/usr/bin/gcc: 5.4.0</strong>
/usr/bin/ruby: N/A
</pre>

[Over at my tap](https://github.com/nwhetsell/homebrew-csound/runs/6434615244?check_suite_focus=true#step:6:35), there’s:

<pre>
CPU: dual-core 64-bit skylake
<strong>Clang: 11.0.0</strong>
Git: 2.36.1 => /usr/bin/git
Curl: 7.68.0 => /usr/bin/curl
Kernel: Linux 5.13.0-1022-azure x86_64 GNU/Linux
OS: Ubuntu 20.04.4 LTS (focal)
<strong>Host glibc: 2.31</strong>
<strong>/usr/bin/gcc: 9.4.0</strong>
/usr/bin/ruby: 2.7.0
</pre>

So, I’m wondering if Csound is being built with either Clang or GCC 9.4 at my tap (resulting in a build that doesn’t segfault), but GCC 5.4 here (resulting in a build that segfaults). If this seems like a possibility, I’m not sure what to do to make Csound use Clang or a more recent GCC on Linux (short of `depends_on "clang"` or `depends_on "gcc"`).

**Comment by SMillerDev** [comment]
Yeah, the homebrew container is not the latest Ubuntu version. You can add depends_on gcc if it's otherwise broken.

**Comment by nwhetsell** [comment]
After adding `depends_on "gcc" => :build`, Csound still segfaults on start-up on Linux here: https://github.com/Homebrew/homebrew-core/runs/6442965111?check_suite_focus=true#step:7:72

This might be the end of the road for this PR, unfortunately.

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:57]:
```suggestion
    depends_on "gcc"
```

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:59]:
```suggestion
fails_with gcc: "5"
```

File: Formula/csound.rb
Comment (danielnachun) [reply Formula/csound.rb:57]: You also need to add `fails_with gcc: "5"` after this.  With those two changes this PR works when I test it locally in our Docker image.

**Comment by danielnachun** [comment]
This last commit has fixed the failure on Linux (!) but now we have what seems to be a similar problem on older macOS.  I am wondering if we can fix this by changing the test to use `.wav` on Linux and `.aif` on macOS.  That's the only thing I can see here that is being changed for macOS.

**Comment by nwhetsell** [comment]
@danielnachun It’s worth a try. Thank you for your help on this!

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:138]: Is there a reason this is now macOS-only?  This doesn't seem like a platform-specific argument like the others are.

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:87]:
```suggestion
    ENV["JAVA_HOME"] = Language::Java.overridable_java_home_env
```
This occurs commonly enough in formulae that we have a dedicated method for it.

**Comment by danielnachun** [comment]
That last fix did the trick!  Once we've addressed the minor changes I've requested we should be ready to go.

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:8]: Are we actually changing anything about how this build on macOS?  If not, I don't think we need to revision bump this.  I'll just dispatch a bottle build once we merge the changes.

File: Formula/csound.rb
Comment (nwhetsell) [reply Formula/csound.rb:138]: * P5 Glove opcodes [depend on a libp5glove library](https://github.com/csound/plugins/blob/a5fd7f194e14833c2bb4f574963613537d6460cb/p5glove/CMakeLists.txt#L5) that’s not generally available on Linux (upstream does something different to build these opcode on macOS). This needn’t be platform specific, but I thought it would be clearer to explicitly indicate that P5 Glove opcodes aren’t being built on Linux.
* Wii remote opcodes depend on [WiiUse](https://github.com/wiiuse/wiiuse), which [does not currently have a Linux bottle](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/wiiuse.rb); this case is similar to the P5 Glove opcodes.
* For the last two, by default Csound installs frameworks and includes in non-standard locations on macOS, so when building Csound plug-ins on macOS these locations need to defined.

Comment (danielnachun) [reply Formula/csound.rb:138]: Ah sorry I should have specified that I was only asking about `CSOUND_INCLUDE_DIR`.  It sounds like it doesn't need to be specified on Linux because it already gets installed there.

Eventually we'll probably want to get `libp5glove` and `wiiuse` bottled on Linux, but since we don't currently have bottles there is no reason we have to wait to bottle this, especially given that they are optional.

Comment (nwhetsell) [reply Formula/csound.rb:138]: No worries!

Just want to add: `include/"csound"` is a symlink to `frameworks/"CsoundLib64.framework/Headers"`:

https://github.com/Homebrew/homebrew-core/blob/7a13dfdee9cbb5548924e247fa741bcaa944cd7e/Formula/csound.rb#L103

The last 3 CMake arguments could be—

```rb
args << "-DCSOUND_FRAMEWORK=#{frameworks}/CsoundLib64.framework"
args << "-DCSOUND_INCLUDE_DIR=#{frameworks}/CsoundLib64.framework/Headers"
args << "-DPLUGIN_INSTALL_DIR=#{frameworks}/CsoundLib64.framework/Resources/Opcodes64"
```

—which might clarify what’s happening.

File: Formula/csound.rb
Comment (nwhetsell) [reply Formula/csound.rb]: Thanks, this is now done.

File: Formula/csound.rb
Comment (nwhetsell) [reply Formula/csound.rb]: We’re not; I reverted the revision to 5.

**Comment by nwhetsell** [comment]
It looks like:

```rb
ENV["JAVA_HOME"] = Language::Java.overridable_java_home_env
```

should be either:

```rb
ENV["JAVA_HOME"] = Language::Java.java_home
```

or:

```rb
ENV << Language::Java.overridable_java_home_env
```

I’m not sure which would be better here.

**Comment by danielnachun** [comment]
Yeah sorry I was slightly wrong - I think `ENV["JAVA_HOME"] = Language::Java.java_home` is preferable here.

File: Formula/csound.rb
Comment (danielnachun) [inline Formula/csound.rb:175]: Sorry to nitpick here, but is there a directory in `HOMEBREW_PREFIX` that users can link the java bindings to that will be picked up as a Java extension?  Or does our OpenJDK hardcode that directory as the only place to find extensions?

Comment (nwhetsell) [reply Formula/csound.rb:175]: > …is there a directory in HOMEBREW_PREFIX that users can link the java bindings to that will be picked up as a Java extension?

Not that I know of.

> …does our OpenJDK hardcode that directory as the only place to find extensions?

As far as I know, Java looks for extensions in the directories defined by its `java.library.path` setting. When I run

```sh
$(brew --prefix openjdk)/bin/java -XshowSettings:properties
```

in a Terminal on Ubuntu 22.04, I see that the value of `java.library.path` is

```
/usr/java/packages/lib
/usr/lib64
/lib64
/lib
/usr/lib
```

The first path seemed like the best choice.

This caveat is based on the Subversion formula (200,000+ installs per year):

https://github.com/Homebrew/homebrew-core/blob/afb158ad34ebd8966dd25ce21bc3a130f485bcef/Formula/subversion.rb#L145-L147

Comment (danielnachun) [reply Formula/csound.rb:175]: It looks like on Linux, by default `java.library.path` is `/usr/java/packages/lib:$LD_LIBRARY_PATH`, so the user could specify the path to the `libexec` folder with `LD_LIBRARY_PATH`, although we don't like to encourage the use of the variable as it can break things.  It looks like it may be possible to pass `-Djava.library.path` when calling the JVM as well.

@Homebrew/core what do we think about this caveat on Linux? Should we at some point try to tweak OpenJDK so it can search some place like `HOMEBREW_PREFIX/lib/java/packages` that any formula can link into.  I don't want to hold up this PR any longer than necessary but I think this is worth getting feedback on.

Comment (nwhetsell) [reply Formula/csound.rb:175]: > …the user could specify the path to the `libexec` folder with `LD_LIBRARY_PATH`, although we don't like to encourage the use of the variable as it can break things.

I figure I should mention that using (what I believe to be) a similar environment variable on macOS (`DYLD_FRAMEWORK_PATH`) was removed from this very formula [relatively recently](https://github.com/Homebrew/homebrew-core/pull/89990).

> …it may be possible to pass `-Djava.library.path` when calling the JVM…

It is; that’s what’s happening in the tests:

https://github.com/Homebrew/homebrew-core/blob/eafc0d300778825bd2b3e7003af91ec136fcf4fa/Formula/csound.rb#L247-L248

Comment (danielnachun) [reply Formula/csound.rb:175]: I'm sufficiently convinced that this is not the result of some underlying issue with our OpenJDK setup and since there has been no other feedback I'm going to merge this and get `csound` bottled.  This caveat can be changed in a syntax-only PR later if needed.

**Review by danielnachun — APPROVED** [review]

**Pull request merged**
